bool second_flag
bool optimization_flag